### PR TITLE
[Gui] Fix color bar position and text

### PR DESCRIPTION
**Pull request by Syres916**
Fixes https://github.com/FreeCAD/FreeCAD/issues/10439 and https://github.com/FreeCAD/FreeCAD/issues/10437

The color bar labels have never been a user configurable setting and on the standard new user background not very easy to read. This PR intends to fix that with improved default settings as well as giving the user the ability to change as they wish. There was a slight change to the top and bottom limits of the size to accommodate the Navigation Cube.

Note: The color bar is not just used in the FEM workbench (the Mesh>CurvatureInfo uses it) hence the settings are in the Display>Colors tab

See https://github.com/FreeCAD/FreeCAD/issues/10437#issuecomment-1702495061 for example screenshot.

### CODE REVIEW

**Comment by sliptonic** [comment]
Is this ready to go?   Not clear from the issue discussion whether this is finished.  If it's not done, please set the PR to draft.

**Comment by Syres916** [comment]
It's ready as far as I'm concerned, sorry I didn't make it clear.

**Comment by Syres916** [comment]
@0penBrain could I please ask for your input on this PR, I genuinely don't understand what the holdup with getting this merged. The FEM crew are really looking forward to getting a decent colour bar y axis.

File: src/Gui/SoFCColorBar.cpp [hunk 1]
```diff
@@ -123,15 +123,15 @@ float SoFCColorBarBase::getBounds(const SbVec2s& size, float& fMinX, float&fMinY
     // The cam height is set in SoFCColorBarBase::getBoundingWidth to 10.
     // Therefore the normalized coordinates are in the range [-5, +5] x [-5ratio, +5ratio] if ratio > 1
     //  and [-5ratio, +5ratio] x [-5, +5] if ratio < 1.
-    // We don't want the whole height covered by the color bar (to have e.g space to the axis cross)
-    // thus we take as base 4.
-    float baseYValue = 4.0f;
+    // We don't want the whole height covered by the color bar (to have e.g space to the axis cross
+    // and the Navigation Cube) thus we take as base 3.25.
+    float baseYValue = 3.25f;
```
Comment (0penBrain) [inline src/Gui/SoFCColorBar.cpp:128]: Works well on "standard case" -- Full HD screen, FC maximized -- but overlay with NaviCube as soon as viewport is reduced (for ex. open the ReportView or minimize FC window).
Wonder how it looks on HiDPI screens.

File: src/Gui/SoFCColorGradient.cpp [hunk 2]
```diff
@@ -99,20 +101,30 @@ void SoFCColorGradient::setMarkerLabel(const SoMFString& label)
         SbVec2f minPt = _bbox.getMin();
         float fStep = (maxPt[1] - minPt[1]) / ((float)num - 1);
         auto trans = new SoTransform;
+
+        ParameterGrp::handle hGrp = Gui::WindowParameter::getDefaultParameter()->GetGroup("View");
+        int LabelTextSize = hGrp->GetInt("CbLabelTextSize", 13);
```
Comment (0penBrain) [inline src/Gui/SoFCColorGradient.cpp:106]: Could use `auto`

File: src/Gui/SoFCColorGradient.cpp [hunk 3]
```diff
@@ -99,20 +101,30 @@ void SoFCColorGradient::setMarkerLabel(const SoMFString& label)
         SbVec2f minPt = _bbox.getMin();
         float fStep = (maxPt[1] - minPt[1]) / ((float)num - 1);
         auto trans = new SoTransform;
+
+        ParameterGrp::handle hGrp = Gui::WindowParameter::getDefaultParameter()->GetGroup("View");
+        int LabelTextSize = hGrp->GetInt("CbLabelTextSize", 13);
+        App::Color LabelTextColor =
```
Comment (0penBrain) [inline src/Gui/SoFCColorGradient.cpp:107]: Could use `auto`

File: src/Gui/SoFCColorGradient.cpp [hunk 4]
```diff
@@ -99,20 +101,30 @@ void SoFCColorGradient::setMarkerLabel(const SoMFString& label)
...
             auto trans = new SoTransform;
             auto color = new SoBaseColor;
             auto text2 = new SoColorBarLabel;
 
             trans->translation.setValue(0, -fStep, 0);
-            color->rgb.setValue(0, 0, 0);
+            color->rgb.setValue(LabelTextColor.r,LabelTextColor.g,LabelTextColor.b);
             text2->string.setValue(label[i]);
             labels->addChild(trans);
             labels->addChild(color);
```
Comment (0penBrain) [inline src/Gui/SoFCColorGradient.cpp:125]: As `color` isn't varying from one label to the other, there is no reason it is added multiple time as child. It should be added only once before the `for` loop.

File: src/Gui/SoFCColorGradient.cpp [hunk 5]
```diff
@@ -99,20 +101,30 @@ void SoFCColorGradient::setMarkerLabel(const SoMFString& label)
...
             labels->addChild(text2);
+            labels->addChild(textFont);
```
Comment (0penBrain) [inline src/Gui/SoFCColorGradient.cpp:127]: Same as `color`. Should be added only once before the `for` loop.

**Comment by 0penBrain** [comment]
> @0penBrain could I please ask for your input on this PR, I genuinely don't understand what the holdup with getting this merged. The FEM crew are really looking forward to getting a decent colour bar y axis.

Well, done. Not sure I did what you expected. :smile: 
Also notice as an issue that the new color/size aren't took into account on existing legends. :confused: In FEM for example, you have to manually recompute the pipeline results.

**Comment by Syres916** [comment]
> 
> In FEM for example, you have to manually recompute the pipeline results.
>

Thanks for feedback, corrections made as required. The only difference I'm seeing is for example opening _data/examples/FemCalculixCantilever3D.FCStd_ I don't have to do any recompute to get the same results but with the correct font and colour, the project just opens as I would expect.


```
OS: Linux Mint 20.3 (X-Cinnamon/cinnamon)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.34292 (Git)
Build type: Release
Branch: master
Hash: 216cb39a762c38f7163f3dc6a5c22dd814247b3a
Python 3.8.10, Qt 5.12.8, Coin 4.0.0, Vtk 7.1.1, OCC 7.6.3
Locale: English/United Kingdom (en_GB)
Installed mods: 
  * Silk 0.1.3
  * Defeaturing 1.2.0
  * CurvedShapes 1.0.4
  * freecad.gears 1.0.0
  * Curves 0.6.13
  * Help 1.0.3
  * fasteners 0.4.56
  * sheetmetal 0.3.0
```

**Comment by 0penBrain** [comment]
> The only difference I'm seeing is for example opening _data/examples/FemCalculixCantilever3D.FCStd_ I don't have to do any recompute to get the same results but with the correct font and colour

Yes, if you change the settings before opening the doc it's OK. My point was with doc already open. In this case a setting change isn't took into account.
This said, fine for me to open a ticket for that point and merge as is.

**Comment by Syres916** [comment]
Ah, I now understand your point, apologies. Yes the two Tree View colors for Objects being edited and Active Containers also behave in this manner so perhaps attack all at once in another PR based on a new issue.

**Comment by Syres916** [comment]
In conclusion this PR is now ready for merge.

**Comment by FEA-eng** [comment]
I've finally managed to build FreeCAD from the source code (on virtual Ubuntu machine) and I tested this PR - it's working correctly:

color bar

Considering this, what @0penBrain said after reviewing and that @Syres916 made the requested changes and concluded that the PR is ready, can I ask you @chennes to merge this?

**Comment by chennes** [comment]
Yes, thanks for the review, @FEA-eng.

**Comment by FEA-eng** [comment]
@Syres916 I'm testing this on a different computer with Windows (new weekly build) now and I've noticed 2 small remaining issues:
- some screen resolutions still lead to the legend partially covered by the navigation cube:



- increasing the label font size may cause the labels to go partially beyond the screen:



But this really applies only to some smaller screen resolutions like on my laptop. Apart from that, it works great.

Would you be able to fix those minor issues in another PR ? If it's too hard to avoid those cases then it's not a big deal.

P.S. A switch from scientific to fixed (decimal) notation would also be awesome and maybe could sometimes help shorten the labels that go outside the screen. But I guess that I should create a separate issue for that.

**Comment by Syres916** [comment]
@FEA-eng I'll have a go at getting a 'best fits' for 1024x768 up to 2560x1440 but I can only test on Linux, The bar may look a little squashed compared to now but I'll post a series of screenshots in my own branch so you can get an idea of now and after the change is applied.

**Comment by Syres916** [comment]
@0penBrain the ticket in question is https://github.com/FreeCAD/FreeCAD/issues/10717 I hope I've made myself more clear in it's description.

**Comment by adrianinsaval** [comment]
is this something we can/should backport?

**Pull request merged**